{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "NGS_Datasets_Chr22.ipynb",
      "provenance": [],
      "collapsed_sections": [
        "jQJMJawH6Ng7"
      ],
      "authorship_tag": "ABX9TyMY5l+3FIWKmak/LGZOd8rY",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/AndreMacedo88/Genomics-Notebooks/blob/main/NGS_Datasets_Chr22.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FfPJjb3TEzHH"
      },
      "source": [
        "# Setting up the machine"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jQJMJawH6Ng7"
      },
      "source": [
        "## Pre-work"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "W1XnOTlX32YP"
      },
      "source": [
        "Instructions for the pre-work got from: http://www.htslib.org/workflow/#mapping_to_variant\n",
        "\n",
        "From the sequencer we get the reads in fastq files, which must be mapped to a genome reference."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Qn2dJUlF6lh4"
      },
      "source": [
        "### Mapping"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "s0SxiImB6oyS"
      },
      "source": [
        "\n",
        "To prepare the reference for mapping you must first index it by typing the following command where <ref.fa> is the path to your reference file:\n",
        "\n",
        "```\n",
        "bwa index <ref.fa>\n",
        "```\n",
        "\n",
        "This may take several hours as it prepares the Burrows Wheeler Transform index for the reference, allowing the aligner to locate where your reads map within that reference.\n",
        "\n",
        "Once you have finished preparing your indexed reference you can map your reads to the reference:\n",
        "\n",
        "```\n",
        "bwa mem -R '@RG\\tID:foo\\tSM:bar\\tLB:library1' <ref.fa> <read1.fa> <read1.fa> > lane.sam\n",
        "```\n",
        "\n",
        "Typically your reads will be supplied to you in two files written in the FASTQ format. It is particularly important to ensure that the @RG information here is correct as this information is used by later tools. The SM field must be set to the name of the sample being processed, and LB field to the library. The resulting mapped reads will be delivered to you in a mapping format known as SAM.\n",
        "\n",
        "Because BWA can sometimes leave unusual FLAG information on SAM records, it is helpful when working with many tools to first clean up read pairing information and flags:\n",
        "\n",
        "```\n",
        "samtools fixmate -O bam <lane.sam> <lane_fixmate.bam>\n",
        "```\n",
        "\n",
        "To sort them from name order into coordinate order:\n",
        "\n",
        "```\n",
        "samtools sort -O bam -o <lane_sorted.bam> -T </tmp/lane_temp> <lane_fixmate.sam>\n",
        "```"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bs4f-_Zu6pyQ"
      },
      "source": [
        "### Improving"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "V1HTtu5_67OC"
      },
      "source": [
        "In order to reduce the number of miscalls of INDELs in your data it is helpful to realign your raw gapped alignment with the Broad’s GATK Realigner.\n",
        "\n",
        "```\n",
        "java -Xmx2g -jar GenomeAnalysisTK.jar -T RealignerTargetCreator -R <ref.fa> -I <lane.bam> -o <lane.intervals> --known <bundle/b38/Mills1000G.b38.vcf>\n",
        "java -Xmx4g -jar GenomeAnalysisTK.jar -T IndelRealigner -R <ref.fa> -I <lane.bam> -targetIntervals <lane.intervals> --known <bundle/b38/Mills1000G.b38.vcf> -o <lane_realigned.bam>\n",
        "```\n",
        "\n",
        "BQSR from the Broad’s GATK allows you to reduce the effects of analysis artefacts produced by your sequencing machines. It does this in two steps, the first analyses your data to detect covariates and the second compensates for those covariates by adjusting quality scores.\n",
        "\n",
        "```\n",
        "java -Xmx4g -jar GenomeAnalysisTK.jar -T BaseRecalibrator -R <ref.fa> -knownSites >bundle/b38/dbsnp_142.b38.vcf> -I <lane.bam> -o <lane_recal.table>\n",
        "java -Xmx2g -jar GenomeAnalysisTK.jar -T PrintReads -R <ref.fa> -I <lane.bam> --BSQR <lane_recal.table> -o <lane_recal.bam>\n",
        "```\n",
        "\n",
        "It is helpful at this point to compile all of the reads from each library together into one BAM, which can be done at the same time as marking PCR and optical duplicates. To identify duplicates we currently recommend the use of either the Picard or biobambam’s mark duplicates tool.\n",
        "\n",
        "```\n",
        "java -Xmx2g -jar MarkDuplicates.jar VALIDATION_STRINGENCY=LENIENT INPUT=<lane_1.bam> INPUT=<lane_2.bam> INPUT=<lane_3.bam> OUTPUT=<library.bam>\n",
        "```\n",
        "\n",
        "Once this is done you can perform another merge step to produce your sample BAM files.\n",
        "```\n",
        "samtools merge <sample.bam> <library1.bam> <library2.bam> <library3.bam>\n",
        "samtools index <sample.bam>\n",
        "```\n",
        "\n",
        "If you have the computational time and resources available it is helpful to realign your INDELS again:\n",
        "\n",
        "```\n",
        "java -Xmx2g -jar GenomeAnalysisTK.jar -T RealignerTargetCreator -R <ref.fa> -I <sample.bam> -o <sample.intervals> --known >bundle/b38/Mills1000G.b38.vcf>\n",
        "java -Xmx4g -jar GenomeAnalysisTK.jar -T IndelRealigner -R <ref.fa> -I <sample.bam> -targetIntervals <sample.intervals> --known >bundle/b38/Mills1000G.b38.vcf> -o <sample_realigned.bam>\n",
        "```\n",
        "\n",
        "Lastly we index our BAM using samtools:\n",
        "\n",
        "```\n",
        "samtools index <sample_realigned.bam>\n",
        "```"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KEvFRASI9cBh"
      },
      "source": [
        "### Variant calling"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "W9ji5NNF9-Px"
      },
      "source": [
        "Variant Calling\n",
        "To convert your BAM file into genomic positions we first use mpileup to produce a BCF file that contains all of the locations in the genome. We use this information to call genotypes and reduce our list of sites to those found to be variant by passing this file into bcftools call.\n",
        "\n",
        "You can do this using a pipe as shown here:\n",
        "\n",
        "```\n",
        "bcftools mpileup -Ou -f <ref.fa> <sample1.bam> <sample2.bam> <sample3.bam> | bcftools call -vmO z -o <study.vcf.gz>\n",
        "```\n",
        "\n",
        "Alternatively if you need to see why a specific site was not called by examining the BCF, or wish to spread the load slightly you can break it down into two steps as follows:\n",
        "\n",
        "```\n",
        "bcftools mpileup -Ob -o <study.bcf> -f <ref.fa> <sample1.bam> <sample2.bam> <sample3.bam>\n",
        "bcftools call -vmO z -o <study.vcf.gz> <study.bcf>\n",
        "```\n",
        "\n",
        "To prepare our VCF for querying we next index it using tabix:\n",
        "\n",
        "```\n",
        "tabix -p vcf <study.vcf.gz>\n",
        "```\n",
        "\n",
        "Additionally you may find it helpful to prepare graphs and statistics to assist you in filtering your variants:\n",
        "\n",
        "```\n",
        "bcftools stats -F <ref.fa> -s - <study.vcf.gz> > <study.vcf.gz.stats>\n",
        "mkdir plots\n",
        "plot-vcfstats -p plots/ <study.vcf.gz.stats>\n",
        "```\n",
        "\n",
        "Finally you will probably need to filter your data using commands such as:\n",
        "\n",
        "```\n",
        "bcftools filter -O z -o <study_filtered..vcf.gz> -s LOWQUAL -i'%QUAL>10' <study.vcf.gz>\n",
        "```\n",
        "\n",
        "Variant filtration is a subject worthy of an article in itself and the exact filters you will need to use will depend on the purpose of your study and quality and depth of the data used to call the variants."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gPj566oNwJMx"
      },
      "source": [
        "## Setup"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "44vBl-WpbJ7y"
      },
      "source": [
        "First need to install some tools like tabix and bgzip, and to import the main python packages"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WuRIcfQKv39y"
      },
      "source": [
        "# This is  used to time the running of this notebook\n",
        "import time\n",
        "start_time = time.time()"
      ],
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WXV5J9VI_dGt",
        "outputId": "0936d137-0dd3-4518-fc83-fc80cf6dbc3e",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 332
        }
      },
      "source": [
        "# check if runtime is clean\n",
        "!ls\n",
        "# get the htslib tools\n",
        "!wget -c https://github.com/samtools/htslib/releases/download/1.11/htslib-1.11.tar.bz2"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "sample_data\n",
            "--2020-10-09 13:00:28--  https://github.com/samtools/htslib/releases/download/1.11/htslib-1.11.tar.bz2\n",
            "Resolving github.com (github.com)... 52.192.72.89\n",
            "Connecting to github.com (github.com)|52.192.72.89|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/4339773/5bc1ce80-fcf0-11ea-8ec0-9a6f894b8e1c?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20201009%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20201009T130029Z&X-Amz-Expires=300&X-Amz-Signature=d5beb0783dd52f9ddd369b31d6d8f7890bf12d5ce3a210ebd16669230dbf6965&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=4339773&response-content-disposition=attachment%3B%20filename%3Dhtslib-1.11.tar.bz2&response-content-type=application%2Foctet-stream [following]\n",
            "--2020-10-09 13:00:29--  https://github-production-release-asset-2e65be.s3.amazonaws.com/4339773/5bc1ce80-fcf0-11ea-8ec0-9a6f894b8e1c?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20201009%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20201009T130029Z&X-Amz-Expires=300&X-Amz-Signature=d5beb0783dd52f9ddd369b31d6d8f7890bf12d5ce3a210ebd16669230dbf6965&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=4339773&response-content-disposition=attachment%3B%20filename%3Dhtslib-1.11.tar.bz2&response-content-type=application%2Foctet-stream\n",
            "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.228.248\n",
            "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.228.248|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1333118 (1.3M) [application/octet-stream]\n",
            "Saving to: ‘htslib-1.11.tar.bz2’\n",
            "\n",
            "htslib-1.11.tar.bz2 100%[===================>]   1.27M  1.19MB/s    in 1.1s    \n",
            "\n",
            "2020-10-09 13:00:31 (1.19 MB/s) - ‘htslib-1.11.tar.bz2’ saved [1333118/1333118]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "9fceOO0SApIT",
        "outputId": "31930bca-1c5a-454b-ea89-23e75e8beda0",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "source": [
        "# install the tools\n",
        "!tar -xjf htslib-1.11.tar.bz2\n",
        "%cd htslib-1.11\n",
        "!./configure --prefix=/htslib\n",
        "!make\n",
        "!make install\n",
        "%cd"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/content/htslib-1.11\n",
            "checking for gcc... gcc\n",
            "checking whether the C compiler works... yes\n",
            "checking for C compiler default output file name... a.out\n",
            "checking for suffix of executables... \n",
            "checking whether we are cross compiling... no\n",
            "checking for suffix of object files... o\n",
            "checking whether we are using the GNU C compiler... yes\n",
            "checking whether gcc accepts -g... yes\n",
            "checking for gcc option to accept ISO C89... none needed\n",
            "checking for ranlib... ranlib\n",
            "checking for grep that handles long lines and -e... /bin/grep\n",
            "checking for C compiler warning flags... -Wall\n",
            "checking for pkg-config... /usr/bin/pkg-config\n",
            "checking pkg-config is at least version 0.9.0... yes\n",
            "checking for special C compiler options needed for large files... no\n",
            "checking for _FILE_OFFSET_BITS value needed for large files... no\n",
            "checking shared library type for unknown-Linux... plain .so\n",
            "checking whether the compiler accepts -fvisibility=hidden... yes\n",
            "checking how to run the C preprocessor... gcc -E\n",
            "checking for egrep... /bin/grep -E\n",
            "checking for ANSI C header files... yes\n",
            "checking for sys/types.h... yes\n",
            "checking for sys/stat.h... yes\n",
            "checking for stdlib.h... yes\n",
            "checking for string.h... yes\n",
            "checking for memory.h... yes\n",
            "checking for strings.h... yes\n",
            "checking for inttypes.h... yes\n",
            "checking for stdint.h... yes\n",
            "checking for unistd.h... yes\n",
            "checking for stdlib.h... (cached) yes\n",
            "checking for unistd.h... (cached) yes\n",
            "checking for sys/param.h... yes\n",
            "checking for getpagesize... yes\n",
            "checking for working mmap... yes\n",
            "checking for gmtime_r... yes\n",
            "checking for fsync... yes\n",
            "checking for drand48... yes\n",
            "checking for srand48_deterministic... no\n",
            "checking whether fdatasync is declared... yes\n",
            "checking for fdatasync... yes\n",
            "checking for library containing log... -lm\n",
            "checking for zlib.h... yes\n",
            "checking for inflate in -lz... yes\n",
            "checking for library containing recv... none required\n",
            "checking for bzlib.h... yes\n",
            "checking for BZ2_bzBuffToBuffCompress in -lbz2... yes\n",
            "checking for lzma.h... yes\n",
            "checking for lzma_easy_buffer_encode in -llzma... yes\n",
            "checking for libdeflate.h... no\n",
            "checking for libdeflate_deflate_compress in -ldeflate... no\n",
            "checking for curl_easy_pause in -lcurl... yes\n",
            "checking for CCHmac... no\n",
            "checking for library containing HMAC... -lcrypto\n",
            "checking whether PTHREAD_MUTEX_RECURSIVE is declared... yes\n",
            "configure: creating ./config.status\n",
            "config.status: creating config.mk\n",
            "config.status: creating htslib.pc.tmp\n",
            "config.status: creating config.h\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o kfunc.o kfunc.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o knetfile.o knetfile.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o kstring.o kstring.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o bcf_sr_sort.o bcf_sr_sort.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o bgzf.o bgzf.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o errmod.o errmod.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o faidx.o faidx.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o header.o header.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o hfile.o hfile.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o hfile_net.o hfile_net.c\n",
            "echo '#define HTS_VERSION_TEXT \"1.11\"' > version.h\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o hts.o hts.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o hts_os.o hts_os.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o md5.o md5.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o multipart.o multipart.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o probaln.o probaln.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o realn.o realn.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o regidx.o regidx.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o region.o region.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o sam.o sam.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o synced_bcf_reader.o synced_bcf_reader.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o vcf_sweep.o vcf_sweep.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o tbx.o tbx.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o textutils.o textutils.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o thread_pool.o thread_pool.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o vcf.o vcf.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o vcfutils.o vcfutils.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/cram_codecs.o cram/cram_codecs.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/cram_decode.o cram/cram_decode.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/cram_encode.o cram/cram_encode.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/cram_external.o cram/cram_external.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/cram_index.o cram/cram_index.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/cram_io.o cram/cram_io.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/cram_samtools.o cram/cram_samtools.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/cram_stats.o cram/cram_stats.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/mFILE.o cram/mFILE.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/open_trace_file.o cram/open_trace_file.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/pooled_alloc.o cram/pooled_alloc.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/rANS_static.o cram/rANS_static.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o cram/string_alloc.o cram/string_alloc.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o hfile_libcurl.o hfile_libcurl.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o hfile_gcs.o hfile_gcs.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o hfile_s3.o hfile_s3.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o hfile_s3_write.o hfile_s3_write.c\n",
            "ar -rc libhts.a kfunc.o knetfile.o kstring.o bcf_sr_sort.o bgzf.o errmod.o faidx.o header.o hfile.o hfile_net.o hts.o hts_os.o md5.o multipart.o probaln.o realn.o regidx.o region.o sam.o synced_bcf_reader.o vcf_sweep.o tbx.o textutils.o thread_pool.o vcf.o vcfutils.o cram/cram_codecs.o cram/cram_decode.o cram/cram_encode.o cram/cram_external.o cram/cram_index.o cram/cram_io.o cram/cram_samtools.o cram/cram_stats.o cram/mFILE.o cram/open_trace_file.o cram/pooled_alloc.o cram/rANS_static.o cram/string_alloc.o   hfile_libcurl.o hfile_gcs.o hfile_s3.o hfile_s3_write.o\n",
            "ranlib libhts.a\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o kfunc.pico kfunc.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o knetfile.pico knetfile.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o kstring.pico kstring.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o bcf_sr_sort.pico bcf_sr_sort.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o bgzf.pico bgzf.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o errmod.pico errmod.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o faidx.pico faidx.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o header.pico header.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o hfile.pico hfile.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o hfile_net.pico hfile_net.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o hts.pico hts.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o hts_os.pico hts_os.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o md5.pico md5.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o multipart.pico multipart.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o probaln.pico probaln.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o realn.pico realn.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o regidx.pico regidx.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o region.pico region.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o sam.pico sam.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o synced_bcf_reader.pico synced_bcf_reader.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o vcf_sweep.pico vcf_sweep.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o tbx.pico tbx.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o textutils.pico textutils.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o thread_pool.pico thread_pool.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o vcf.pico vcf.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o vcfutils.pico vcfutils.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/cram_codecs.pico cram/cram_codecs.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/cram_decode.pico cram/cram_decode.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/cram_encode.pico cram/cram_encode.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/cram_external.pico cram/cram_external.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/cram_index.pico cram/cram_index.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/cram_io.pico cram/cram_io.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/cram_samtools.pico cram/cram_samtools.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/cram_stats.pico cram/cram_stats.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/mFILE.pico cram/mFILE.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/open_trace_file.pico cram/open_trace_file.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/pooled_alloc.pico cram/pooled_alloc.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/rANS_static.pico cram/rANS_static.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o cram/string_alloc.pico cram/string_alloc.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o hfile_libcurl.pico hfile_libcurl.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o hfile_gcs.pico hfile_gcs.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o hfile_s3.pico hfile_s3.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -fpic -c -o hfile_s3_write.pico hfile_s3_write.c\n",
            "gcc -shared -Wl,-soname,libhts.so.3 -fvisibility=hidden  -o libhts.so kfunc.pico knetfile.pico kstring.pico bcf_sr_sort.pico bgzf.pico errmod.pico faidx.pico header.pico hfile.pico hfile_net.pico hts.pico hts_os.pico md5.pico multipart.pico probaln.pico realn.pico regidx.pico region.pico sam.pico synced_bcf_reader.pico vcf_sweep.pico tbx.pico textutils.pico thread_pool.pico vcf.pico vcfutils.pico cram/cram_codecs.pico cram/cram_decode.pico cram/cram_encode.pico cram/cram_external.pico cram/cram_index.pico cram/cram_io.pico cram/cram_samtools.pico cram/cram_stats.pico cram/mFILE.pico cram/open_trace_file.pico cram/pooled_alloc.pico cram/rANS_static.pico cram/string_alloc.pico hfile_libcurl.pico hfile_gcs.pico hfile_s3.pico hfile_s3_write.pico -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "ln -sf libhts.so libhts.so.3\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o bgzip.o bgzip.c\n",
            "gcc -fvisibility=hidden  -o bgzip bgzip.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o htsfile.o htsfile.c\n",
            "gcc -fvisibility=hidden  -o htsfile htsfile.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o tabix.o tabix.c\n",
            "gcc -fvisibility=hidden  -o tabix tabix.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/hts_endian.o test/hts_endian.c\n",
            "gcc -fvisibility=hidden  -o test/hts_endian test/hts_endian.o -llzma -lbz2 -lz -lm   -lcurl -lcrypto\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/fieldarith.o test/fieldarith.c\n",
            "gcc -fvisibility=hidden  -o test/fieldarith test/fieldarith.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/hfile.o test/hfile.c\n",
            "gcc -fvisibility=hidden  -o test/hfile test/hfile.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/pileup.o test/pileup.c\n",
            "gcc -fvisibility=hidden  -o test/pileup test/pileup.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/plugins-dlhts.o test/plugins-dlhts.c\n",
            "gcc -fvisibility=hidden  -o test/plugins-dlhts test/plugins-dlhts.o -llzma -lbz2 -lz -lm   -lcurl -lcrypto\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/sam.o test/sam.c\n",
            "gcc -fvisibility=hidden  -o test/sam test/sam.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test_bgzf.o test/test_bgzf.c\n",
            "gcc -fvisibility=hidden  -o test/test_bgzf test/test_bgzf.o libhts.a -lz -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test_kfunc.o test/test_kfunc.c\n",
            "gcc -fvisibility=hidden  -o test/test_kfunc test/test_kfunc.o libhts.a -lz -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test_kstring.o test/test_kstring.c\n",
            "gcc -fvisibility=hidden  -o test/test_kstring test/test_kstring.o libhts.a -lz -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test_realn.o test/test_realn.c\n",
            "gcc -fvisibility=hidden  -o test/test_realn test/test_realn.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test-regidx.o test/test-regidx.c\n",
            "gcc -fvisibility=hidden  -o test/test-regidx test/test-regidx.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test_str2int.o test/test_str2int.c\n",
            "gcc -fvisibility=hidden  -o test/test_str2int test/test_str2int.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test_view.o test/test_view.c\n",
            "gcc -fvisibility=hidden  -o test/test_view test/test_view.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test_index.o test/test_index.c\n",
            "gcc -fvisibility=hidden  -o test/test_index test/test_index.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test-vcf-api.o test/test-vcf-api.c\n",
            "gcc -fvisibility=hidden  -o test/test-vcf-api test/test-vcf-api.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test-vcf-sweep.o test/test-vcf-sweep.c\n",
            "gcc -fvisibility=hidden  -o test/test-vcf-sweep test/test-vcf-sweep.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test-bcf-sr.o test/test-bcf-sr.c\n",
            "gcc -fvisibility=hidden  -o test/test-bcf-sr test/test-bcf-sr.o libhts.a -lz -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/fuzz/hts_open_fuzzer.o test/fuzz/hts_open_fuzzer.c\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test-bcf-translate.o test/test-bcf-translate.c\n",
            "gcc -fvisibility=hidden  -o test/test-bcf-translate test/test-bcf-translate.o libhts.a -lz -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "gcc -Wall -g -O2 -fvisibility=hidden -I.  -c -o test/test-parse-reg.o test/test-parse-reg.c\n",
            "gcc -fvisibility=hidden  -o test/test-parse-reg test/test-parse-reg.o libhts.a -llzma -lbz2 -lz -lm   -lcurl -lcrypto -lpthread\n",
            "sed -n '/^static_libs=/s/[^=]*=/HTSLIB_static_LIBS = /p;/^static_ldflags=/s/[^=]*=/HTSLIB_static_LDFLAGS = /p' htslib.pc.tmp > htslib_static.mk\n",
            "sed -e 's#@-includedir@#'`pwd`'#g;s#@-libdir@#'`pwd`'#g' htslib.pc.tmp > htslib-uninstalled.pc\n",
            "mkdir -p -m 755 /htslib/bin /htslib/include /htslib/include/htslib /htslib/lib /htslib/share/man/man1 /htslib/share/man/man5 /htslib/share/man/man7 /htslib/lib/pkgconfig\n",
            "if test -n \"\"; then mkdir -p -m 755 ; fi\n",
            "install -p -m 644 libhts.so /htslib/lib/libhts.so.1.11\n",
            "ln -sf libhts.so.1.11 /htslib/lib/libhts.so\n",
            "ln -sf libhts.so.1.11 /htslib/lib/libhts.so.3\n",
            "sed -e 's#@-includedir@#/htslib/include#g;s#@-libdir@#/htslib/lib#g;s#@-PACKAGE_VERSION@#1.11#g' htslib.pc.tmp > /htslib/lib/pkgconfig/htslib.pc\n",
            "chmod 644 /htslib/lib/pkgconfig/htslib.pc\n",
            "install -p bgzip htsfile tabix /htslib/bin\n",
            "if test -n \"\"; then install -p  ; fi\n",
            "install -p -m 644 htslib/*.h /htslib/include/htslib\n",
            "install -p -m 644 libhts.a /htslib/lib/libhts.a\n",
            "install -p -m 644 bgzip.1 htsfile.1 tabix.1 /htslib/share/man/man1\n",
            "install -p -m 644 faidx.5 sam.5 vcf.5 /htslib/share/man/man5\n",
            "install -p -m 644 htslib-s3-plugin.7 /htslib/share/man/man7\n",
            "/root\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "2KfO6_exE7sn"
      },
      "source": [
        "# add the tools to PATH\n",
        "import os\n",
        "os.environ[\"PATH\"] += os.pathsep + \"/htslib/bin\"\n",
        "# test if tabix and bgzig are installed correctly\n",
        "!tabix"
      ],
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ZNSj3vPpwe46"
      },
      "source": [
        "### Install Python packages"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZoWa2oxNloX1",
        "outputId": "fa195cfe-a7cb-41fb-c56b-c93b8b1b5b91",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 468
        }
      },
      "source": [
        "%%time\n",
        "# Install general packages\n",
        "!pip install matplotlib\n",
        "!pip install numpy\n",
        "\n",
        "!pip install biopython\n",
        "!pip install PyVCF"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.6/dist-packages (3.2.2)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib) (2.4.7)\n",
            "Requirement already satisfied: numpy>=1.11 in /usr/local/lib/python3.6/dist-packages (from matplotlib) (1.18.5)\n",
            "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib) (2.8.1)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib) (1.2.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.6/dist-packages (from matplotlib) (0.10.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.6/dist-packages (from python-dateutil>=2.1->matplotlib) (1.15.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.6/dist-packages (1.18.5)\n",
            "Collecting biopython\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/76/02/8b606c4aa92ff61b5eda71d23b499ab1de57d5e818be33f77b01a6f435a8/biopython-1.78-cp36-cp36m-manylinux1_x86_64.whl (2.3MB)\n",
            "\u001b[K     |████████████████████████████████| 2.3MB 2.8MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.6/dist-packages (from biopython) (1.18.5)\n",
            "Installing collected packages: biopython\n",
            "Successfully installed biopython-1.78\n",
            "Collecting PyVCF\n",
            "  Downloading https://files.pythonhosted.org/packages/20/b6/36bfb1760f6983788d916096193fc14c83cce512c7787c93380e09458c09/PyVCF-0.6.8.tar.gz\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.6/dist-packages (from PyVCF) (50.3.0)\n",
            "Building wheels for collected packages: PyVCF\n",
            "  Building wheel for PyVCF (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for PyVCF: filename=PyVCF-0.6.8-cp36-cp36m-linux_x86_64.whl size=121984 sha256=0162d358fbd8d9487448c2416f0c1f7e306405a5edbef61fb7a20e9389c0179e\n",
            "  Stored in directory: /root/.cache/pip/wheels/81/91/41/3272543c0b9c61da9c525f24ee35bae6fe8f60d4858c66805d\n",
            "Successfully built PyVCF\n",
            "Installing collected packages: PyVCF\n",
            "Successfully installed PyVCF-0.6.8\n",
            "CPU times: user 64.9 ms, sys: 29 ms, total: 93.9 ms\n",
            "Wall time: 16 s\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kCDBD30uEayz"
      },
      "source": [
        "# Analyzing variant calls"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tbJpr117bw1t"
      },
      "source": [
        "After running the Pre-work, or another genotype caller (for example, GATK or samtools mpileup), you will have a Variant Call Format (VCF) file reporting on genomic variations, such as SNPs (Single-Nucleotide Polymorphisms), InDels (Insertions/Deletions), CNVs (Copy Number Variation) among others."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TsuxzW72XHYG",
        "outputId": "e18f8b5f-a404-4737-8fac-c062ad2cda00",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 52
        }
      },
      "source": [
        "# perform a partial download of the VCF file for chromosome 22 (up to 17 Mbp) of the 1000 genomes project. Then, bgzip will compress it.\n",
        "!tabix -fh ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/vcf_with_sample_level_annotation/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5_extra_anno.20130502.genotypes.vcf.gz 22:1-17000000|bgzip -c > genotypes.vcf.gz\n",
        "# create a tabix index file for the bgzip-compressed VCF, which we will need for direct access to a section of the genome.\n",
        "!tabix -p vcf genotypes.vcf.gz\n",
        "# The tabix command appends .tbi to the .vcf.gz filename, creating a binary index file named .vcf.gz.tbi with which genomic coordinates can quickly be translated into file offsets in .vcf.gz."
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[E::easy_errno] Libcurl reported error 78 (Remote file not found)\n",
            "[E::easy_errno] Libcurl reported error 78 (Remote file not found)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GTOWgpn9eaB7",
        "outputId": "bcc7d495-1a18-4a10-dcf8-69e50b085ad9",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 416
        }
      },
      "source": [
        "# check if files correctly downloaded and processed\n",
        "!ls\n",
        "!gzip -dk genotypes.vcf.gz\n",
        "!head -20 genotypes.vcf\n",
        "!rm genotypes.vcf"
      ],
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "ALL.chr22.phase3_shapeit2_mvncall_integrated_v5_extra_anno.20130502.genotypes.vcf.gz.tbi\n",
            "genotypes.vcf.gz\n",
            "genotypes.vcf.gz.tbi\n",
            "##fileformat=VCFv4.1\n",
            "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
            "##fileDate=20140730\n",
            "##reference=ftp://ftp.1000genomes.ebi.ac.uk//vol1/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz\n",
            "##source=1000GenomesPhase3Pipeline\n",
            "##contig=<ID=1,assembly=b37,length=249250621>\n",
            "##contig=<ID=2,assembly=b37,length=243199373>\n",
            "##contig=<ID=3,assembly=b37,length=198022430>\n",
            "##contig=<ID=4,assembly=b37,length=191154276>\n",
            "##contig=<ID=5,assembly=b37,length=180915260>\n",
            "##contig=<ID=6,assembly=b37,length=171115067>\n",
            "##contig=<ID=7,assembly=b37,length=159138663>\n",
            "##contig=<ID=8,assembly=b37,length=146364022>\n",
            "##contig=<ID=9,assembly=b37,length=141213431>\n",
            "##contig=<ID=10,assembly=b37,length=135534747>\n",
            "##contig=<ID=11,assembly=b37,length=135006516>\n",
            "##contig=<ID=12,assembly=b37,length=133851895>\n",
            "##contig=<ID=13,assembly=b37,length=115169878>\n",
            "##contig=<ID=14,assembly=b37,length=107349540>\n",
            "##contig=<ID=15,assembly=b37,length=102531392>\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-Eb5FwSCErdE"
      },
      "source": [
        "# Final consideration"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "DcTfCXud6JSi",
        "outputId": "705854d1-1f24-4ff9-fa55-e900d002b549",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        }
      },
      "source": [
        "# Running time of the notebook\n",
        "print(\"{:.2f} minutes\".format((time.time()-start_time)/60))"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "2.08 minutes\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}